packages/replay/src/coreHandlers/util/fetchUtils.ts

```diff
@@ -0,0 +1,140 @@
+  getBodySize,
+  getBodyString,
+  getNetworkBody,
+  makeNetworkReplayBreadcrumb,
+  parseContentLengthHeader,
+} from './networkUtils';
+
+/**
+ * Capture a fetch breadcrumb to a replay.
+ * This adds additional data (where approriate).
+ */
+export async function captureFetchBreadcrumbToReplay(
+  breadcrumb: Breadcrumb & { data: FetchBreadcrumbData },
+  hint: FetchHint,
+  options: { captureBodies: boolean; textEncoder: TextEncoderInternal; replay: ReplayContainer },
+): Promise<void> {
+  try {
+    const data = await _prepareFetchData(breadcrumb, hint, options);
+
+    // Create a replay performance entry from this breadcrumb
+    const result = makeNetworkReplayBreadcrumb('resource.fetch', data);
+    addNetworkBreadcrumb(options.replay, result);
+  } catch (error) {
+    __DEBUG_BUILD__ && logger.error('[Replay] Failed to capture fetch breadcrumb', error);
+  }
+}
+
+/**
+ * Enrich a breadcrumb with additional data.
+ * This has to be sync & mutate the given breadcrumb,
+ * as the breadcrumb is afterwards consumed by other handlers.
+ */
+export function enrichFetchBreadcrumb(
+  breadcrumb: Breadcrumb & { data: FetchBreadcrumbData },
+  hint: FetchHint,
+  options: { textEncoder: TextEncoderInternal },
+): void {
+  const { input, response } = hint;
+
+  const body = _getFetchRequestArgBody(input);
+  const reqSize = getBodySize(body, options.textEncoder);
+  const resSize = response ? parseContentLengthHeader(response.headers.get('content-length')) : undefined;
+
+  if (reqSize !== undefined) {
+    breadcrumb.data.request_body_size = reqSize;
+  }
+  if (resSize !== undefined) {
+    breadcrumb.data.response_body_size = resSize;
+  }
+}
+
+async function _prepareFetchData(
+  breadcrumb: Breadcrumb & { data: FetchBreadcrumbData },
+  hint: FetchHint,
+  options: { captureBodies: boolean; textEncoder: TextEncoderInternal },
+): Promise<ReplayNetworkRequestData> {
+  const { startTimestamp, endTimestamp } = hint;
+
+  const { url, method, status_code: statusCode, request_body_size: requestBodySize } = breadcrumb.data;
+  let responseBodySize = breadcrumb.data.response_body_size;
+  let responseBody: NetworkBody | undefined;
+  let requestBody: NetworkBody | undefined;
+
+  // Only clone the response if we need to
+  if (options.captureBodies || responseBodySize === undefined) {
+    try {
+      // We have to clone this, as the body can only be read once
+      const response = hint.response.clone();
+      const { body, bodyText } = await _parseFetchBody(response);
+
+      if (bodyText && bodyText.length && responseBodySize === undefined) {
+        responseBodySize = getBodySize(bodyText, options.textEncoder);
+      }
+
+      if (options.captureBodies && responseBodySize && responseBodySize < NETWORK_BODY_MAX_SIZE) {
+        responseBody = body;
+      }
+    } catch {
+      // just ignore if something fails here
+    }
+  }
+
+  if (options.captureBodies && requestBodySize && requestBodySize < NETWORK_BODY_MAX_SIZE) {
```

Comment:
I assume request bodies will generally be much smaller than response, we may want a different limit on request vs response.

Reply:
IMHO this limit should be the "hard" limit, informed by what we def. don't want. In the future we'll probably prefer some kind of truncation first anyhow, so hopefully few things will fall into this at the end!